async_engine.cpp

```diff
@@ -0,0 +1,92 @@
+#include "async_engine.hpp"
+#include "socket_exception.hpp"
+#include <iostream>
+#include <sys/select.h>
+
+AsyncEngine::AsyncEngine(Socket_sptr master_socket) : m_master_socket(master_socket)
+{
+    FD_ZERO(&m_fd_set_read);
+    FD_SET(*master_socket, &m_fd_set_read);
+    m_max_fd = *master_socket;
+}
+
+AsyncEngine::~AsyncEngine()
+{
+    try
+    {
+        m_master_socket.reset();
+    }
+    catch (SocketException & se)
+    {
+        std::cout << "~AsyncEngine() failed with error: " << se.what() << std::endl;
+    }
+}
+
+Socket_vector AsyncEngine::Select(const std::chrono::milliseconds & timeout)
+{
+    fd_set temp_set;
+    memcpy(&temp_set, &m_fd_set_read, sizeof(m_fd_set_read));
+    size_t nfds = m_max_fd + 1;
+    timeval dt;
+    std::chrono::seconds full_seconds = std::chrono::duration_cast<std::chrono::seconds>(timeout);
+    std::chrono::microseconds useconds = timeout - full_seconds;
+    dt.tv_sec = full_seconds.count();
+    dt.tv_usec = useconds.count();
+    int ready_sockets_count = select(nfds, &temp_set,  NULL, NULL, &dt);
+    if(ready_sockets_count < 0)
+        throw SocketException("select() failed");
+
+    Socket_vector ready_sockets;
+
+    if(FD_ISSET(*m_master_socket, &temp_set))
+        ready_sockets.push_back(m_master_socket);
+
+    for (auto it = m_accepted_sockets.begin(); it != m_accepted_sockets.end(); ++it)
+    {
+        if(FD_ISSET(*(it->get()), &temp_set))
```

Comment:
You can get rid of this weird construction using range-based `for` loop.
But note: such loop creates copies of container elements. To avoid copying use `auto&`.